### PR TITLE
Store comparator hits in the CLCT (CCLUT-4)

### PR DESCRIPTION
**Pull request by dildick**
#### PR description:

This PR saves the comparator hits that were used to build a CLCT in a ComparatorContainer. (Similar to how a CSCSegment has access to the CSCRecHit2Ds.) The relevant change is in function `patternFinding`, which called for both the trigger and the pre-trigger.  This function now also identifies all hits that were part of a particular pattern. An intermediate pattern map keeps track of all comparator hits, for each key half-strip and for each pattern Id. Ultimately, only the comparator hits for the best pattern are stored in the CLCT. 

No changes to the actual trigger logic have been made. No changes are expected in the output. 

#### PR validation:

Tested with WF 22034.0.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

N/A

@tahuang1991

### CODE REVIEW

**Comment by dildick** [comment]
Hi, can someone initiate the tests please?

**Comment by smuzaffar** [comment]
please test

**Comment by dildick** [comment]
There are some failures in the L1T and L1TEMU directories, none in the CSC directory. Not sure where those failures are coming from, but they are not related to this PR.

**Comment by dildick** [comment]
@rekovic Do you have any more questions on this pull request?

**Comment by silviodonato** [comment]
do you have other comments @rekovic  ?

**Comment by silviodonato** [comment]
@rekovic ?

File: L1Trigger/CSCTriggerPrimitives/src/CSCCathodeLCTProcessor.cc [hunk 1]
```diff
@@ -832,15 +867,26 @@ bool CSCCathodeLCTProcessor::patternFinding(const PulseArray pulse, const int nS
...
+      // clear all layers
+      for (int ilayer = 0; ilayer < CSCConstants::NUM_LAYERS; ilayer++) {
         hit_layer[ilayer] = false;
+      }
+
+      // clear a single pattern!
+      CSCCLCTDigi::ComparatorContainer hits_single_pattern;
+      hits_single_pattern.resize(6);
+      for (auto& p : hits_single_pattern) {
+        p.resize(CSCConstants::CLCT_PATTERN_WIDTH, 65535);
```
Comment (rekovic) [inline L1Trigger/CSCTriggerPrimitives/src/CSCCathodeLCTProcessor.cc:886]: Magic number 65535, please store as a named constant.

File: L1Trigger/CSCTriggerPrimitives/src/CSCCathodeLCTProcessor.cc [hunk 2]
```diff
@@ -938,6 +998,12 @@ void CSCCathodeLCTProcessor::markBusyKeys(const int best_hstrip,
   }
 }  // markBusyKeys -- TMB-07 version.
 
+void CSCCathodeLCTProcessor::cleanComparatorContainer(CSCCLCTDigi::ComparatorContainer& compHits) const {
+  for (auto& p : compHits) {
+    p.erase(std::remove_if(p.begin(), p.end(), [](unsigned i) -> bool { return i == 65535; }), p.end());
```
Comment (rekovic) [inline L1Trigger/CSCTriggerPrimitives/src/CSCCathodeLCTProcessor.cc:1003]: Magic number 65535, please store as a named constant.

**Comment by dildick** [comment]
@rekovic I added an enum in the processor code.

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Pull request merged**